metricbeat/module/kubernetes/container/data.go

```diff
@@ -97,29 +95,33 @@ func eventMapping(content []byte, perfMetrics *util.PerfMetricsCache) ([]common.
 				},
 			}
 
-			if nodeCores > 0 {
-				containerEvent.Put("cpu.usage.node.pct", float64(container.CPU.UsageNanoCores)/1e9/nodeCores)
-			}
-
-			if nodeMem > 0 {
-				containerEvent.Put("memory.usage.node.pct", float64(container.Memory.UsageBytes)/nodeMem)
-			}
-
-			cuid := util.ContainerUID(pod.PodRef.Namespace, pod.PodRef.Name, container.Name)
-			coresLimit := perfMetrics.ContainerCoresLimit.GetWithDefault(cuid, nodeCores)
-			memLimit := perfMetrics.ContainerMemLimit.GetWithDefault(cuid, nodeMem)
-
-			if coresLimit > 0 {
-				containerEvent.Put("cpu.usage.limit.pct", float64(container.CPU.UsageNanoCores)/1e9/coresLimit)
-			}
-
-			if memLimit > 0 {
-				containerEvent.Put("memory.usage.limit.pct", float64(container.Memory.UsageBytes)/memLimit)
-			}
+			containers = append(containers, containerEvent)
+		}
+	}
 
-			events = append(events, containerEvent)
+	events := util.MergeEvents(containers, stateMetrics,
+		map[string]string{
+			mb.ModuleDataKey + ".node.name": node.NodeName,
+		},
```

Comment:
I guess the `/metrics` endpoint doesn't offer any filtering... There can be a huge list here with big clusters.

Reply:
Yes, I have some concerns about this approach, I'm trying something else, let's keep this PR unmerged in the meanwhile.

Thanks for the review!

Reply:
Hi, let me chime in because I'm being affected by this just now 😄 

----

Yes, so my best bet was to "silently" run `state_*` metricsets to populate the `perfMetrics` only for consumption by the `pod` metricset. In this mode, `state_*` metricsets shouldn't publish anything:

```yaml
- module: kubernetes
  metricsets:
    - pod
    - container
    - node
  hosts: ["192.168.99.100:10255"]  # kubelet
  period: 10s
  state_metrics:
    host: "kube-state-metrics:8080"
```

And `state_*` metricsets are enabled for publishing only when you explicitly listed them under the `metricsets`:

```yaml
- module: kubernetes
  metricsets:
    - state_container
    - state_node
  hosts: ["192.168.99.100:10255"]  # kubelet
  period: 10s
  state_metrics:
    host: "kube-state-metrics:8080"
```

Then you can deploy the former one in a k8s daemonset, whereas the latter is deployed in a k8s deployment.

But obviously my suggested implementation doesn't resolve #6637.

Reply:
#7470 seems awesome! But not sure what you mean by the perfMetrics part yet.

Would you be enabling an enhanced version of `NodeMetadataEnricher` on the container metricset [here](https://github.com/elastic/beats/pull/7470/files#diff-96e4742da1cc1cb927b8fdd795b09e39R70), in addition to the ContainerMetadataEnricher, in order to fill `perfMetrics.NodeCoresAllocatable` that is used for calculating `container.cpu.usage.node.pct`?

Thanks anyway for your efforts as always! I'll keep eyes on #7470 👍

Reply:
yes, that's the plan :smiley_cat: